components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/obsrvability/handler/metrics/publisher/MetricHandler.java

```diff
@@ -0,0 +1,480 @@
+                String name = synCtx.getProperty(MetricConstants.PROXY_NAME).toString();
+                incrementProxyErrorCount(name);
+                stopTimers(synCtx.getProperty(MetricConstants.PROXY_LATENCY_TIMER), synCtx);
+            } else if ((null != synCtx.getProperty(SynapseConstants.IS_INBOUND) &&
+                    synCtx.getProperty(SynapseConstants.IS_INBOUND).toString().equals("true")) ||
+                    ((null != axis2MessageContext.getProperty("TransportInURL")) &&
+                            !axis2MessageContext.getProperty("TransportInURL").toString().contains("services"))) {
+                // Get the port used in service invoking
+                if (null != ((Axis2MessageContext) synCtx).getAxis2MessageContext().
+                        getProperty(NhttpConstants.SERVICE_PREFIX)) {
+                    String servicePort = ((Axis2MessageContext) synCtx).getAxis2MessageContext().
+                            getProperty(NhttpConstants.SERVICE_PREFIX).toString();
+                    servicePort = servicePort.substring((servicePort.lastIndexOf(':') + 1),
+                            servicePort.lastIndexOf(DELIMITER));
+                    serviceInvokePort = Integer.parseInt(servicePort);
+                }
+                if ((null != synCtx.getProperty(SynapseConstants.IS_INBOUND))) {
+                    String inboundEndpointName = synCtx.getProperty(MetricConstants.INBOUND_ENDPOINT_NAME).
+                            toString();
+                    incrementInboundEndpointErrorCount(inboundEndpointName);
+                    stopTimers(synCtx.getProperty
+                            (MetricConstants.INBOUND_ENDPOINT_LATENCY_TIMER), synCtx);
+                } else if (null != synCtx.getProperty(MetricConstants.SYNAPSE_REST_API) &&
+                        (serviceInvokePort != internalHttpApiPort)) {
+                    String context = axis2MessageContext.getProperty(MetricConstants.TRANSPORT_IN_URL).
+                            toString();
+                    String apiInvocationUrl = axis2MessageContext.getProperty(MetricConstants.SERVICE_PREFIX).
+                            toString() + context.replaceFirst(DELIMITER, EMPTY);
+                    String apiName = getApiName(context, synCtx);
+
+                    incrementApiErrorCount(apiName, apiInvocationUrl);
+                    stopTimers(synCtx.getProperty(MetricConstants.API_LATENCY_TIMER), synCtx);
+                }
+            }
+        }
+        synCtx.setProperty(SynapseConstants.IS_ERROR_COUNT_ALREADY_PROCESSED, true);
+        return true;
+    }
+
+    @Override
+    public boolean handleStopServer() {
+        String host = System.getProperty(MicroIntegratorBaseConstants.LOCAL_IP_ADDRESS);
+        String port = System.getProperty(MetricConstants.HTTP_PORT);
+        String javaVersion = System.getProperty(MetricConstants.JAVA_VERSION);
+        String javaHome = System.getProperty(MetricConstants.JAVA_HOME);
```

Comment:
this code segment is repeated in init as well .. we need to read the system properties only once and use it. I guess no need to read it again, wdyt

Reply:
Moved them to the class scope.